### PR TITLE
*: support user defined filters for baseline capture

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/15655

Problem Summary:

New functionality for baseline capture.

### What is changed and how it works?

What's Changed:
- add a new system table `mysql.capture_plan_baselines_blacklist`
- users can insert / update / delete rows in `mysql.capture_plan_baselines_blacklist` to manipulate the capture filter
- the table contains 3 columns, `id`, `filter_type`, and `filter_value`, currently, only `table` / `db` / `frequency` type values are supported, other unknown types would be ignored with a warning message in the log
- `table` type values should be like `test.t`, i.e, database name should be specified for the table name, otherwise, it would be ignored with a warning in the log
-  if `frequency` type value is 3, then all the queries which occur less than 4 times would be ignored in the capture; if there are multiple `frequency` rows, the capture filter would choose the largest one
- there is no cache for the rows in `mysql.capture_plan_baselines_blacklist`, so no reload operation is needed once the table data is modified; the baseline capture would read the table each time

### Check List 

Tests 

- [x] Unit test

Side effects

N/A

Documentation

- [x] Affects user behaviors

### Release note 

- Support user defined filters for baseline capture

### CODE REVIEW

File: bindinfo/bind_test.go [hunk 1]
```diff
@@ -2114,3 +2114,157 @@ func (s *testSuite) TestTemporaryTable(c *C) {
...
+	tk.MustExec("admin capture bindings")
+	rows = tk.MustQuery("show global bindings").Rows()
+	c.Assert(len(rows), Equals, 1)
+	c.Assert(rows[0][0], Equals, "select * from `test` . `t` where `a` > ?")
+	tk.MustExec("delete from mysql.capture_plan_baselines_blacklist")
+
+	// Case sensitivity.
+	s.cleanBindingEnv(tk)
+	stmtsummary.StmtSummaryByDigestMap.Clear()
+	tk.MustExec("insert into mysql.capture_plan_baselines_blacklist(filter_type, filter_value) values('tABle', 'tESt.T')")
```
Comment (Reminiscent) [inline bindinfo/bind_test.go:2243]: Should we handle the situation that the string from the `filter_value` contains the space?

File: bindinfo/handle.go [hunk 2]
```diff
@@ -609,10 +610,104 @@ func (c cache) getBindRecord(hash, normdOrigSQL, db string) *BindRecord {
...
+	// No need to acquire the session context lock for ExecRestrictedStmt, it
+	// uses another background session.
+	rows, _, err := exec.ExecRestrictedStmt(context.TODO(), stmt)
+	if err != nil {
+		logutil.BgLogger().Warn("[sql-bind] failed to load mysql.capture_plan_baselines_blacklist", zap.Error(err))
+		return
+	}
+	for _, row := range rows {
+		filterTp := strings.ToLower(row.GetString(0))
+		valStr := strings.ToLower(row.GetString(1))
```
Comment (Reminiscent) [inline bindinfo/handle.go:670]: Should we standardize the string when we store it?

File: bindinfo/handle.go [hunk 3]
```diff
@@ -609,10 +610,104 @@ func (c cache) getBindRecord(hash, normdOrigSQL, db string) *BindRecord {
...
+		switch filterTp {
+		case "db":
+			filter.dbs[valStr] = struct{}{}
+		case "table":
+			strs := strings.Split(valStr, ".")
```
Comment (Reminiscent) [inline bindinfo/handle.go:675]: Should we handle more situations like there are some spaces in the string?

**Review by Reminiscent — APPROVED** [review]

File: bindinfo/handle.go (hunk 2 above)
Comment (eurekaka) [reply bindinfo/handle.go:670]: Then we have to hijack the normal `insert` / `update` statements, to specially handle the manipulations on this table, which is pretty tricky.

File: bindinfo/handle.go (hunk 3 above)
Comment (eurekaka) [reply bindinfo/handle.go:675]: We can do it later if users really insert stupid records like that.

File: bindinfo/handle.go [hunk 4]
```diff
@@ -653,10 +654,104 @@ func (c cache) getBindRecord(hash, normdOrigSQL, db string) *BindRecord {
...
+			f, err := strconv.ParseInt(valStr, 10, 64)
+			if err != nil {
+				logutil.BgLogger().Warn("[sql-bind] failed to parse frequency type value, ignore it", zap.String("filter_value", valStr), zap.Error(err))
+				continue
+			}
+			if f < 1 {
+				logutil.BgLogger().Warn("[sql-bind] frequency threshold is less than 1, ignore it", zap.Int64("frequency", f))
+				continue
+			}
+			if f > filter.frequency {
```
Comment (winoros) [inline bindinfo/handle.go:739]: So this means that the value of the `frequency` is maintained by the user. We just pick the biggest one they insert.

Comment (eurekaka) [reply bindinfo/handle.go:739]: Yes.

**Review by winoros — APPROVED** [review]

**Pull request merged**